pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java

```diff
@@ -143,30 +143,13 @@ public TimestampUtils(boolean usesDouble, Provider<TimeZone> timeZoneProvider) {
     this.timeZoneProvider = timeZoneProvider;
   }
 
-  private Calendar getCalendar(int sign, int hr, int min, int sec) {
-    int rawOffset = sign * (((hr * 60 + min) * 60 + sec) * 1000);
-    if (calCache != null && calCacheZone == rawOffset) {
+  private Calendar getCalendar(ZoneOffset offset) {
+    if (calCache != null && Objects.equals(offset, calCacheZone)) {
       return calCache;
     }
 
-    StringBuilder zoneID = new StringBuilder("GMT");
-    zoneID.append(sign < 0 ? '-' : '+');
-    if (hr < 10) {
-      zoneID.append('0');
-    }
-    zoneID.append(hr);
-    if (min < 10) {
-      zoneID.append('0');
-    }
-    zoneID.append(min);
-    if (sec < 10) {
-      zoneID.append('0');
-    }
-    zoneID.append(sec);
-
-    TimeZone syntheticTZ = new SimpleTimeZone(rawOffset, zoneID.toString());
-    calCache = new GregorianCalendar(syntheticTZ);
-    calCacheZone = rawOffset;
+    calCache = new GregorianCalendar(TimeZone.getTimeZone(offset));
+    calCacheZone = offset;
```

Comment:
I wonder if this change has performance implications.
`ZoneOffset` should be fine as it stores the offset via `int totalSeconds`.

@uschindler , have you checked `TimeZone.getTimeZone(offset)` performance?

Reply:
I did not do a performance check, but `TimeZone.getTimeZone(offset)` is not globally synchronized like the one taking a string parameter.

There was a previous commit where this was changed at another place: https://github.com/pgjdbc/pgjdbc/commit/beeabe2ead05e0cc6303acd6e3ff65a26623d2fd

I will come back to this a bit later by checking what going on behind the scenes.

Reply:
I checked the code in `TimeZone.getTimeZone(ZoneOffset)` (Java 8): Basically it is doing more or less the same. It 

If you like, you can replace the code and also use `SimpleTimeZone` with `"GMT" + offset.getId()` and call it like before, but `TimeZone.getTimeZone(ZoneOffset)` has caching for the zone ids, so it looks better to use this code.

At least there's no synchronization involved.

I am not sure whet the best way to check performance is, you are the better specialist!

Reply:
In addition, there's also the cache, so the performance of this method is not significant (unless the strings to be parsed all the time have different time zones). This is unlikely because the backend (Postgres) returns the same zone offsets (server's local one, but only in text-representations). After some time the JDBC driver switches to binary transmission and then this method is no longer called.

So I don't think this is an issue. The cache is helpful to work around the raw-offset to string conversions.

Reply:
If you still feel better, I can change the method to:

```java
  private Calendar getCalendar(ZoneOffset offset) {
    if (calCache != null && Objects.equals(offset, calCacheZone)) {
      return calCache;
    }

    final String tzid = (offset.getTotalSeconds() == 0) ? "UTC" : "GMT".concat(offset.getId());
    final TimeZone syntheticTZ = new SimpleTimeZone(offset.getTotalSeconds() * 1000, tzid);
    calCache = new GregorianCalendar(syntheticTZ);
    calCacheZone = offset;
    return calCache;
  }
```

The `tzid` string operations are needed because the ID of raw offsets are not using "GMT" prefix and in case of UTC, it returns "Z". But actually that doe snot matter, we could also add a constant fake ID, because the ID is never used by Java or Calendar.

Reply:
The conversions of the name were adapted from the conversions original `TimeZone.getTimeZone(ZoneId)` source code.

Reply:
Hi, we figured out that there is some strange difference between `SimpleTimeZone(offset.getTotalSeconds() * 1000)` and `TimeZone.getTimeZone(offset)`.

So we should fix this in #2488 by applying above suggestion. This is still a server-side problem in combination with problematic Java time zone handling for offsets going down to minutes/seconds.

Reply:
Here is the JDK bug report: https://bugs.openjdk.java.net/browse/JDK-8285844